cloudinary-image-management-and-manipulation-in-the-cloud-cdn/php/class-sync.php

```diff
@@ -119,66 +163,654 @@ public function is_synced( $post_id ) {
+				$signature = array();
 			}
+
+			// Remove any old or outdated signature items. against the expected.
+			$signature                    = array_intersect_key( $signature, $this->sync_types );
+			$signatures[ $attachment_id ] = $return;
+			$return                       = wp_parse_args( $signature, $this->sync_types );
 		}
 
 		return $return;
 	}
 
+	/**
+	 * Generate a new Public ID for an asset.
+	 *
+	 * @param int $attachment_id The attachment ID for the new public ID.
+	 *
+	 * @return string|null
+	 */
+	public function generate_public_id( $attachment_id ) {
+
+		$cld_folder = $this->managers['media']->get_cloudinary_folder();
+		$file       = get_attached_file( $attachment_id );
+		$file_info  = pathinfo( $file );
+		$public_id  = $cld_folder . $file_info['filename'];
+
+		return ltrim( $public_id, '/' );
+	}
+
+	/**
+	 * Maybe add a suffix to the public ID if it's not unique.
+	 *
+	 * @param string      $public_id     The public ID to maybe add a suffix.
+	 * @param int         $attachment_id The attachment ID.
+	 * @param string|null $suffix        The suffix to maybe add.
+	 *
+	 * @return string The public ID.
+	 */
+	public function add_suffix_maybe( $public_id, $attachment_id, $suffix = null ) {
+
+		// Test if asset exists by calling just the head on the asset url, to prevent API rate limits.
+		$url         = $this->plugin->components['connect']->api->cloudinary_url( $public_id . $suffix );
+		$req         = wp_remote_head( $url, array( 'body' => array( 'rdm' => wp_rand( 100, 999 ) ) ) );
+		$asset_error = strtolower( wp_remote_retrieve_header( $req, 'x-cld-error' ) );
+		$code        = wp_remote_retrieve_response_code( $req );
+
+		// If the request is not a 404 & does not have a cld-error header stating resource not found, it exists and should be checked that it's not a resync or generate a prefixed ID.
+		if ( 404 !== $code && false === strpos( $asset_error, 'resource not found' ) ) {
+
+			// Get the attachment type.
+			if ( wp_attachment_is( 'image', $attachment_id ) ) {
+				$type = 'image';
+			} elseif ( wp_attachment_is( 'video', $attachment_id ) ) {
+				$type = 'video';
+			} elseif ( wp_attachment_is( 'audio', $attachment_id ) ) {
+				$type = 'audio';
+			} else {
+				// not supported.
+				return null;
+			}
+			$cld_asset = $this->plugin->components['connect']->api->get_asset_details( $public_id, $type );
+			if ( ! is_wp_error( $cld_asset ) && ! empty( $cld_asset['public_id'] ) ) {
+				$context_id = null;
+
+				// Exists, check to see if this asset originally belongs to this ID.
+				if ( ! empty( $cld_asset['context'] ) && ! empty( $cld_asset['context']['custom'] ) && ! empty( $cld_asset['context']['custom']['wp_id'] ) ) {
+					$context_id = (int) $cld_asset['context']['custom']['wp_id'];
+				}
+
+				// Generate new ID only if context ID is not related.
+				if ( $context_id !== $attachment_id ) {
+					// Generate a new ID with a uniqueID prefix.
+					$suffix = '-' . uniqid();
```

Comment:
`uniqid` will not guarantee uniqueness and will fail because of the system's clock:
> This function does not guarantee uniqueness of return value. Since most systems adjust system clock by NTP or like, system time is changed constantly. Therefore, it is possible that this function does not return unique ID for the process/thread.

Providing the `more_entropy` parameter might help, but still doesn't guarantee uniqueness. I suggest we go for a stronger random string here, such as uuid v4 or something else - to ensure we're never generating the same string twice.

Reply:
Agreed, this is largely untouched from the revert last week.
This is what I'm working on today and tomorrow.

Reply:
@dugajean @DavidCramer 

I'm not sure if we aren't overcomplicating. What will be the problem if the combination of the filename and suffix already exists in Cloudinary? The system will find the next available combination.

Reply:
@pereirinha - yes. But I'm reworking this so this isn't even going to be in. The method is unused and never called.